Kilnjs (#1405)
clay/clay-kiln: docs/kilnjs.md

---
id: kilnjs
title: Kilnjs
sidebar_label: Kilnjs
---

---

> Kilnjs is an optional Javascript file that transforms the schema.yaml file into a dynamic Javascript Object that affects how Kiln interacts with components that use the schema.  With it, events can be attached to the inputs that are used by kiln to edit components.  These events can be used to hide or reveal different fields in the schema, update the values of those fields, make api calls, subscribe to Vuex mutations, and just about any other action allowed by JavaScript.

Reviewer: Just a couple of suggestions
```suggestion
Kiln.js is an optional JavaScript file that transforms the `schema.yaml` file into a dynamic JavaScript Object that affects how Kiln interacts with components that use the schema.  With it, events can be attached to the inputs that are used by kiln to edit components. These events can be used to hide or reveal different fields in the schema, update the values of those fields, make API calls, subscribe to Vuex mutations, and just about any other action allowed by JavaScript.
```

Author: 👍

Using Kilnjs, you can remove the logic from the schema.yaml file and move it into JavaScript, keeping the schema file for presentational information.

To take advantage of the power that Kilnjs provides you need to create a file called kiln.js within a component folder, in the same location where you place the component's schema, model, and client files. At a minimum, it should look like this.

> ```

Reviewer 2: Missing language definition for styles and format

```suggestion
```js
```

Author: 👍

'use strict';

module.exports = (schema) => {

  return schema;
};
```

> A function that recieves a parameter called schema that also returns the schema.  The schema contains a JSON representation of the schema.yaml file and can be manipulated before returning.  Values can be changed, added, or deleted directly on the JSON object without any other outside code.  Anything you can do to a JSON object can be done to the schema.

Author: 👍

> A function that recieves a parameter called schema that also returns the schema.  The schema contains a JSON representation of the schema.yaml file and can be manipulated before returning.  Values can be changed, added, or deleted directly on the JSON object without any other outside code.  Anything you can do to a JSON object can be done to the schema.

Reviewer: ```suggestion
A function that receives a parameter called schema that also returns the schema.  The schema contains a JSON representation of the `schema.yaml` file and can be manipulated before returning.  Values can be changed, added, or deleted directly on the JSON object without any other outside code.  Anything you can do to a JSON object can be done to the schema.
```

> ```

Reviewer: Missing language definition for styles and format
```suggestion
```js
```

Author: 👍

schema['_groups'].settings['_placeholder'].height = '200px';
```

> However, note that the JSON that is returned still needs to conform to the structure needed by Kiln.  Deleting properties or adding properties that Kiln doesn't recognize can and will result in errors.

Reviewer 2: should we add ⚠️  signs here?

Author: How do I add those so they show in docusaurus?

Author: 👍

---

## KilnInput

> The real power of Kilnjs comes from the KilnInput object, which can be used to make the fields within a schema truly dynamic.  Using KilnInput you can add events to the different form inputs as detailed on the [Form Inputs page](input).  You can also subscribe to Vuex actions as described below.

Reviewer: ```suggestion
The real power of Kilnjs comes from the KilnInput object, which can be used to make the fields within a schema truly dynamic. Using `KilnInput` you can add events to the different form inputs as detailed on the [Form Inputs page](input). You can also subscribe to Vuex actions as described below.
```

Author: 👍

> To make a schema field a KilnInput, you set it to a new instance of KilnInput, passing it to the schema and the name of the field.

Reviewer: ```suggestion
To make a schema field a `KilnInput`, you set it to a new instance of `KilnInput`, passing it to the schema and the name of the field.
```

Author: 👍

> ```

Reviewer: Remember to add the language in here
```suggestion
```js
```

Author: 👍

const KilnInput = window.kiln.kilnInput;

module.exports = (schema) => {
  schema.title = new KilnInput(schema, 'title');

  return schema;
};
```

---

## Events

> You can attach events to the schema inputs.  The events you can attach vary depending upon the type input. The different events are detailed on the [Form Inputs page](input).

Reviewer 2: ```suggestion
You can attach events to the schema inputs. The events you can attach vary depending upon the type input. The different events are detailed on the [Form Inputs page](input).
```

Author: 👍

> * ***on(event, callbackFunction)*** - pass it the event as a string, along with a callback function that is run after the event happens

Reviewer 2: ```suggestion
* ***on(event, callbackFunction)*** - pass it the event as a string, along with a callback function that is run after the event happens.
```

Author: 👍

```
  schema.enableSocialButtons.on('input', (val) => {

>     if(val) {

Reviewer: ```suggestion
    if (val) {
```
Also, remember to add the language to the code snippet for styles and format

Author: 👍

      schema.shareServices.show();
    } else {
      schema.shareServices.hide();
    }
  });
```

### Validation

By attaching events, one of the things you can use kilnjs for is to do field validation.

> ```

Reviewer 2: ```suggestion
```js
```

Author: 👍

module.exports = (schema) => {
  schema.title = new KilnInput(schema, 'title');

  schema.title.on('keydown', (e) => {
    // prevent numbers from being entered in the title field
    if (!isNaN(e.key)) {
      e.preventDefault();
    }
  });

  return schema;
};
```

---

## Vuex Actions to Subscribe To

> The following are some of the Vuex actions that you can subscribe to using Kiln.js. The actions can be scoped so that they only trigger the subscription function when the action is triggered by a component of the same type as the subscribing component. In other words, if scope is set to true on a paragraph element, then the subscription function would only be called when a paragraph element triggers the Vuex action, but not when any other type of component triggers the Vuex action.  When the scope is set to false, the subscription function is called whenever _any_ component triggers the Vuex action.

Reviewer 3: I think we can add `the` in front of the word `scope` in the sentence; 
`In other words, if scope is set to true on a paragraph element`

Author: 👍

> The following are some of the Vuex actions that you can subscribe to using Kiln.js. The actions can be scoped so that they only trigger the subscription function when the action is triggered by a component of the same type as the subscribing component. In other words, if scope is set to true on a paragraph element, then the subscription function would only be called when a paragraph element triggers the Vuex action, but not when any other type of component triggers the Vuex action.  When the scope is set to false, the subscription function is called whenever _any_ component triggers the Vuex action.

Reviewer: ```suggestion
The following are some of the Vuex actions that you can subscribe to using Kiln.js. The actions can be scoped so that they only trigger the subscription function when the action is triggered by a component of the same type as the subscribing component. In other words, if `scope` is set to true on a paragraph element, then the subscription function would only be called when a paragraph element triggers the Vuex action, but not when any other type of component triggers the Vuex action. When the scope is set to false, the subscription function is called whenever _any_ component triggers the Vuex action.
```

> The following are some of the Vuex actions that you can subscribe to using Kiln.js. The actions can be scoped so that they only trigger the subscription function when the action is triggered by a component of the same type as the subscribing component. In other words, if the scope is set to true on a paragraph element, then the subscription function would only be called when a paragraph element triggers the Vuex action, but not when any other type of component triggers the Vuex action.  When the scope is set to false, the subscription function is called whenever _any_ component triggers the Vuex action.

Reviewer 2: **Suggestion**: Highlight the word `true` and `false`. Also Remove extra space before the word `When`

Author: 👍

> * ***OPEN_FORM*** - Triggered when a Kiln form is opened. Note that this can triggered by both a modal form _and_ an inline form, so just clicking into a paragraph or other inline field will trigger this action. The payload returned from this action contains a JSON representation of the Schema along with the component uri that was opened in the form.

Reviewer: ```suggestion
* ***OPEN_FORM*** - Triggered when a Kiln form is opened. Note that this can be triggered by both a modal form _and_ an inline form, so just clicking into a paragraph or other inline field will trigger this action. The payload returned from this action contains a JSON representation of the Schema along with the component URI that was opened in the form.
```

Author: 👍

> * ***CLOSE_FORM*** - Triggered when a Kiln form is closed. For inline fields this occurs when the field loses focus.  When the user clicks out of the field.  The payload returned from this action is undefined.

Reviewer 3: ```suggestion
* ***CLOSE_FORM*** - Triggered when a Kiln form is closed. For inline fields, this occurs when the field loses focus.  When the user clicks out of the field.  The payload returned from this action is undefined.
```

> * ***CLOSE_FORM*** - Triggered when a Kiln form is closed. For inline fields this occurs when the field loses focus.  When the user clicks out of the field.  The payload returned from this action is undefined.

Reviewer: ```suggestion
* ***CLOSE_FORM*** - Triggered when a Kiln form is closed. For inline fields this occurs when the field loses focus.  When the user clicks out of the field. The payload returned from this action is `undefined`.
```

Author: 👍

> * ***UPDATE_FORMDATA*** - Triggered when a user changes the data in a form field.  This happens "on change", _not_ "on save". The payload returns and object containing the path (field being changed) and data (the new value).

Reviewer 3: I think should be `an` here. What you think
```suggestion
* ***UPDATE_FORMDATA*** - Triggered when a user changes the data in a form field.  This happens "on change", _not_ "on save". The payload returns an object containing the path (field being changed) and data (the new value).
```

Author: user starts witha "Yoo" sound, and y is a consonant, so a is correct.

Reviewer 3: Hello @scottnash 
Yes, that's correct; but I mentioned the `and` before `object` 😅 
Did `suggested change` not show the update?

> * ***UPDATE_FORMDATA*** - Triggered when a user changes the data in a form field.  This happens "on change", _not_ "on save". The payload returns and object containing the path (field being changed) and data (the new value).

Reviewer: ```suggestion
* ***UPDATE_FORMDATA*** - Triggered when a user changes the data in a form field. This happens "on change", _not_ "on save". The payload returns and object containing the path (field being changed) and data (the new value).
```

> * ***UPDATE_FORMDATA*** - Triggered when an user changes the data in a form field.  This happens "on change", _not_ "on save". The payload returns an object containing the path (field being changed) and data (the new value).

Reviewer 2: ```suggestion
* ***UPDATE_FORMDATA*** - Triggered when an user changes the data in a form field. ⚠️This happens "on change", _not_ "on save" ⚠️. The payload returns an object containing the path (field being changed) and data (the new value).
```

Author: 👍

> * ***UPDATE_COMPONENT*** - Triggered when a component is saved.  The payload returns a json object containing all the fields in the component's schema, with the values for each of them.

Reviewer: ```suggestion
* ***UPDATE_COMPONENT*** - Triggered when a component is saved. The payload returns a JSON object containing all the fields in the component's schema, with the values for each of them.
```

Author: 👍

> * ***UPDATE_PAGE_STATE*** - Triggered when the page state changes.  When it is published, unpublished, etc. The payload returned contains the page meta information, including published date/time, update date/time, history, users, etc.

Reviewer: ```suggestion
* ***UPDATE_PAGE_STATE*** - Triggered when the page state changes. When it is published, unpublished, etc. The payload returned contains the page meta information, including published date/time, update date/time, history, users, etc.
```

Author: 👍

> Often you might not wish to associate a subscription with a specific field, but rather to something more general.  You can instantiate a KilnInput without referencing a specific field. Of course, if it makes sense to create a connection between a field and the action involved with a subscription, you can use the subscribe function on a field.  It can be done either way.  It's just a matter of what makes the intent clearer.

Reviewer: ```suggestion
Often you might not wish to associate a subscription with a specific field, but rather to something more general.  You can instantiate a `KilnInput` without referencing a specific field. Of course, if it makes sense to create a connection between a field and the action involved with a subscription, you can use the subscribe function on a field. It can be done either way. It's just a matter of what makes the intent clearer.
```

Author: 👍

```
module.exports = (schema) => {
  const subscriptions = new KilnInput(schema);

  schema.myDate = new KilnInput(schema, 'myDate');

>   subscriptions.subscribe('UPDATE_COMPONENT', (payload)=> {

Reviewer: ```suggestion
  subscriptions.subscribe('UPDATE_COMPONENT', (payload) => {
```

Author: 👍

    // do something when a component has been updated
    // you can also test the type of component updated
    // by checking the value of payload.data.componentVariation
    // and only reacting when it's a certain type of component
    // or if the component's new value is equal or not to some value
  }, false);

>   schema.myDate.subscribe('UPDATE_COMPONENT', (payload)=> {

Reviewer: ```suggestion
  schema.myDate.subscribe('UPDATE_COMPONENT', (payload) => {
```

Author: 👍

    // same as above, just attached to the myDate field, which implies
    // you want to do something to the myDate field with the payload
  });

  return schema;
};

```

---

## Kilnjs custom methods

> KilnInput also provides its own set of custom methods.

Reviewer: ```suggestion
`KilnInput` also provides its own set of custom methods.
```

Author: 👍

### getComponentInstances

* ***getComponentInstances(componentName)*** - returns an array containing the uris of all components of type componentName that are on the current page.

```
kilnInput.getComponentInstances('paragraph');
```

### reRenderInstance

> * ***reRenderInstance(uri)*** - Fetchs the component's data and passes it through the components model.render function, thus refreshing it on the page. For instance, if some outside source has updated component data and does not trigger a page refresh, you could force a component rerender that would include the updated data.

Reviewer: ```suggestion
* ***reRenderInstance(uri)*** - Fetches the component's data and passes it through the components `model.render` function, thus refreshing it on the page. For instance, if some outside source has updated component data and does not trigger a page refresh, you could force a component re-render that would include the updated data.
```

> * ***reRenderInstance(uri)*** - Fetchs the component's data and passes it through the components model.render function, thus refreshing it on the page. For instance, if some outside source has updated component data and does not trigger a page refresh, you could force a component rerender that would include the updated data.

Reviewer 3: Sorry miss this yesterday 😅 
```suggestion
* ***reRenderInstance(uri)*** - Fetches the component's data and passes it through the components model.render function, thus refreshing it on the page. For instance, if some outside source has updated component data and does not trigger a page refresh, you could force a component to re-render that would include the updated data.
```

Author: Thanks!  Corrected

```
// if a paragraph component is updated, then rerender the instances of the snash component

> kilnInput.subscribe('UPDATE_COMPONENT', (payload)=> {

Reviewer: ```suggestion
kilnInput.subscribe('UPDATE_COMPONENT', (payload) => {
```

Author: 👍

    if (payload.data.componentVariation === 'paragraph') {
      let snashInstances = eventBus.getComponentInstances('snash');
      snashInstances.forEach((instance) => {
        kilnInput.reRenderInstance(instance);
      });
    }
  }, false);
```

### saveComponent

> * ***saveComponent(uri, data)*** - Saves the component at the provided uri with the provided data.  Saving a component this way will also rerender the it on the page with the updated data.

Reviewer 3: ```suggestion
* ***saveComponent(uri, data)*** - Saves the component at the provided uri with the provided data.  Saving a component this way will also re-render it on the page with the updated data.
```

Author: 👍

> * ***saveComponent(uri, data)*** - Saves the component at the provided uri with the provided data.  Saving a component this way will also rerender the it on the page with the updated data.

Reviewer: ```suggestion
* ***saveComponent(uri, data)*** - Saves the component at the provided URI with the provided data. Saving a component this way will also re-render the it on the page with the updated data.
```

Author: 👍

```
const uri = 'localhost/_components/paragraph/instances/cjue4dl7i00062a65rpexptmp',
  data = { text: "This is a paragraph" };

kilnInput.saveComponent(uri, data);
```

### setProp

* ***setProp(prop, value)*** - change value of a property on the input

```
kilninput.setProp('_has', { ...kilninput['_has'], input: 'select' });
```

### show/hide

* ***show()*** - used to make an input visible.

* ***hide()*** - used to make an input invisible

```
  schema.enableSocialButtons.on('input', (val) => {

>     if(val) {

Reviewer: ```suggestion
    if (val) {
```

Author: 👍

      schema.shareServices.show();
    } else {
      schema.shareServices.hide();
    }
  });
```

### url

> * ***url()*** - returns the url object from Vuex state of the component that is currently being edited. Object contains the component name, the instance, and the path

Reviewer 3: ```suggestion
* ***url()*** - returns the url object from Vuex state of the component that is currently being edited. The object contains the component name, the instance, and the path
```

Author: 👍

> * ***url()*** - returns the url object from Vuex state of the component that is currently being edited. Object contains the component name, the instance, and the path

Reviewer: Missing dot at the end of this sentence

Author: 👍

```
  kilnInput.url()

  /// returns an object structured like this

  {
    component: "meta-title"
    instance: "cjtfuc3rw00019fz9egagqev0"
    path: "settings"

>   }

Reviewer: I feel that it would be better if we put this whole block inside a comment block
```js
kilnInput.url()

/**
 * returns an object structured like this
 *
 * {
 *   component: "meta-title"
 *   instance: "cjtfuc3rw00019fz9egagqev0"
 *   path: "settings"
 * }
 */
```

Author: 👍

---

> ⚠️You must being using Clay-cli version 3.10.2 or greater to use Kiln.js⚠️

Reviewer 2: Suggestion: we could add bold style to the version

⚠️You must being using Clay-cli version **3.10.2** or greater to use Kiln.js⚠️

Author: 👍
